core/src/main/scala/org/apache/spark/PreemptionPolicy.scala

```diff
@@ -0,0 +1,121 @@
+import org.apache.spark.scheduler.cluster.CoarseGrainedClusterMessages.PreemptExecutors
+import org.apache.spark.util.Utils
+
+/**
+ * PreemptionPolicy companion object abstracts loading the
+ * `DYN_ALLOCATION_PREEMPTION_POLICY`.
+ */
+object PreemptionPolicy {
+  private def getCtor(conf: SparkConf): Constructor[_] = {
+    Utils
+      .classForName(conf.get(DYN_ALLOCATION_PREEMPTION_POLICY))
+      .getConstructor(
+        classOf[mutable.HashSet[String]],
+        classOf[mutable.HashMap[String, Long]],
+        classOf[mutable.HashSet[String]]
+      )
+  }
+
+  def mkPolicy(
+      conf: SparkConf,
+      executorIds: mutable.HashSet[String],
+      removeTimes: mutable.HashMap[String, Long],
+      preemptedExecutors: mutable.HashSet[String]): PreemptionPolicy = {
+    getCtor(conf)
+      .newInstance(executorIds, removeTimes, preemptedExecutors)
+      .asInstanceOf[PreemptionPolicy]
+  }
+}
+
+/**
+ * The preemptExecutors method is synchronized in the ExecutorAllocationManager (EAM),
+ * so it's not necessary here. But still...beware.
+ *
+ * @param executorIds mutable collection of executorIds (not mutated by default)
+ * @param removeTimes EAM's idle executor -> expiration map (not mutated by default)
+ * @param preemptedExecutors mutable collection of executors to be preempted on the next
+ *                           EAM update. preemptExecutors adds to this set, but EAM acts
+ *                           on and clears the set
+ */
+abstract class PreemptionPolicy(
+    executorIds: mutable.HashSet[String],
+    removeTimes: mutable.HashMap[String, Long],
+    preemptedExecutors: mutable.HashSet[String]) {
+  def preemptExecutors(pe: PreemptExecutors): Unit
+}
+
+/**
+ * DefaultPreemptionPolicy is the default preemption policy selected when
+ * DYN_ALLOCATION_PREEMPTION_POLICY is not set. This policy was created with YARN's
+ * PreemptionMessage symantics in mind. Instead of simply killing YARN's requested
+ * containers, preemptively remove idle executors without cached data. If YARN requests
+ * to preempt more executors than are idle, include the `askedToLeave` set and randomly
+ * select executors to fill out the total preemption count.
+ */
+class DefaultPreemptionPolicy(
+    executorIds: mutable.HashSet[String],
+    removeTimes: mutable.HashMap[String, Long],
+    preemptedExecutors: mutable.HashSet[String]) extends
+  PreemptionPolicy(executorIds, removeTimes, preemptedExecutors) {
+
+  protected[spark] def orderByTimeout(
+      execMap: mutable.HashMap[String, Long],
+      numExecs: Int): Seq[String] = {
+    execMap.toSeq.sortBy(_._2).take(numExecs).map(_._1)
+  }
+
+  /**
+   * find executors to offer up for preemption
+   * favor most idle executors without cached data
+   * NOTE: does not consider non-idle executors
+   */
+  protected[spark] def findPreemptableExecutors(numExecsToKill: Int): Seq[String] = {
+    val blockMaster = SparkEnv.get.blockManager.master
```

Comment:
Recommend that you make this logic clearer by extracting a tuple (has cached data, idle time) for each executor, then sorting that list of tuples, then taking first k from the sorted list. Or perhaps just sort all executors on the number of bytes of cached data and kill the first k in that list.